Sources/Datadog/Tracing/TracingAutoInstrumentation.swift

```diff
@@ -0,0 +1,113 @@
+            return nil
+        }
+    }
+
+    func apply() {
+        let interceptor = TracingRequestInterceptor.build(with: tracedHosts)
+        swizzler.swizzle(using: interceptor)
+    }
+}
+
+internal enum TracingRequestInterceptor {
+    static func build(with tracedHosts: Set<URL>) -> RequestInterceptor {
+        let interceptor: RequestInterceptor = { urlRequest in
+            guard let tracer = Global.sharedTracer as? DDTracer,
+                let someURL = urlRequest.url,
+                tracedHosts.allows(someURL),
+                HTTPHeadersWriter.canInject(to: urlRequest) else {
+                    return nil
+            }
+            let spanContext = tracer.createSpanContext()
+            let headersWriter = HTTPHeadersWriter()
+            headersWriter.inject(spanContext: spanContext)
+            let tracingHeaders = headersWriter.tracePropagationHTTPHeaders
+            var modifiedRequest = urlRequest
+            tracingHeaders.forEach { modifiedRequest.setValue($1, forHTTPHeaderField: $0) }
+
+            let observer: TaskObserver = tracingTaskObserver(tracer: tracer, spanContext: spanContext)
+            return InterceptionResult(modifiedRequest: modifiedRequest, taskObserver: observer)
+        }
+        return interceptor
+    }
+
+    static func tracingTaskObserver(
+        tracer: DDTracer,
+        spanContext: DDSpanContext
+    ) -> TaskObserver {
+        var startedSpan: OTSpan? = nil
+        let observer: TaskObserver = { observedEvent in
+            switch observedEvent {
+            case .starting(let request):
+                if let ongoingSpan = startedSpan {
+                    userLogger.warn("\(String(describing: request)) is starting a new trace but it's already started a trace before: \(ongoingSpan)")
+                    developerLogger?.warn("\(String(describing: request)) is starting a new trace but it's already started a trace before: \(ongoingSpan)")
+                }
+                let span = tracer.startSpan(
+                    spanContext: spanContext,
+                    operationName: "urlsession.request"
+                )
+                let url = request?.url?.absoluteString ?? "unknown_url"
+                let method = request?.httpMethod ?? "unknown_method"
+                span.setTag(key: DDTags.resource, value: url)
+                span.setTag(key: OTTags.httpUrl, value: url)
+                span.setTag(key: OTTags.httpMethod, value: method)
+                startedSpan = span
+            case .completed(let response, let error):
+                guard let completedSpan = startedSpan else {
+                    break
+                }
+                if let statusCode = (response as? HTTPURLResponse)?.statusCode {
+                    completedSpan.setTag(key: OTTags.httpStatusCode, value: statusCode)
+                }
+                if let someError = error {
+                    completedSpan.handleError(someError)
```

Comment:
When testing by running the `Example` app on Simulator and browsing seeded traces on app.datadoghq.com I spotted another misalignment with Android SDK. Because the app tries to download data from `https://app.datadoghq.com//dataTaskWithURL`, it results with `HTTP 404` error. Following the [Android SDK implementation](https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt#L78-L93), this should have a separate handling based on the HTTP status codes:
```kotlin
span.setTag(Tags.HTTP_STATUS.key, statusCode)
if (statusCode in 400..499) {
   (span as? MutableSpan)?.isError = true
}
if (statusCode == 404) {
   // limit spam on 404 resources
   (span as? MutableSpan)?.resourceName = "404"
}
```
Such implementation will:
* mark the span with error label in Traces Explorer,
* group all `404s` under single `"404"` RESOURCE facet for easy discovery by users.

Reply:
copied Android's imp ✅